htlcswitch/switch.go

```diff
@@ -0,0 +1,599 @@
+	// which is used to save the payments which made by user.
+	pendingPayments map[[sha256.Size]byte]*pendingPayment
+
+	// circuits is an structure which is used to forward the settle/fail
+	// htlc updates back to the add htlc initiator.
+	circuits *circuitMap
+
+	// links is a map of channel output and channel link which manages
+	// this channel.
+	links map[wire.OutPoint]ChannelLink
+
+	// commands is used to propagate the commands to the handler goroutine.
+	commands chan interface{}
+}
+
+// New creates the new instance of htlc switch.
+func New() *Switch {
+	return &Switch{
+		circuits:        newCircuitMap(),
+		links:           make(map[wire.OutPoint]ChannelLink),
+		pendingPayments: make(map[[sha256.Size]byte]*pendingPayment),
+		commands:        make(chan interface{}),
+		quit:            make(chan bool, 1),
+	}
+}
+
+// SendUpdate is used by other subsystems which aren't belong to htlcswicth
+// package in order to send the payment.
+// TODO(andrew.shvv) lnwire messages should be hidden from upper levels.
+func (s *Switch) SendUpdate(dest routing.HopID, htlc *lnwire.UpdateAddHTLC) (
+	chan [sha256.Size]byte, chan error) {
+
+	// Create payment and add to the map of payment in order later to be
+	// able to retrieve it and return response to the user.
+	payment := &pendingPayment{
+		err:      make(chan error, 1),
+		preimage: make(chan [sha256.Size]byte, 1),
+	}
+	s.pendingPayments[htlc.PaymentHash] = payment
+
+	// Generate and send new update packet, if error will be received
+	// on this stage it means that packet didn't left boundaries of our
+	// system and something wrong happened.
+	var source wire.OutPoint
+	if err := s.Forward(newUpdatePacket(source, dest, htlc)); err != nil {
+		payment.err <- err
+	}
+
+	// Returns channels so that other subsystem might wait/skip the
+	// waiting of handling of payment.
+	return payment.preimage, payment.err
+}
+
+// Forward is used in order to find next channel link and apply htlc
+// update. Also this function is used by channel links itself in order to
+// forward the update after it has been included in the channel.
+func (s *Switch) Forward(packet *htlcPacket) error {
+	command := &forwardPacketCmd{
+		pkt: packet,
+		err: make(chan error),
+	}
+
+	select {
+	case s.commands <- command:
+		return <-command.err
+	case <-s.quit:
+		return errors.New("htlc switch was stopped")
+	}
+}
+
+// handleForward handles incoming forward commands.
+func (s *Switch) handleForward(command *forwardPacketCmd) {
+	var paymentHash [sha256.Size]byte
+
+	switch m := command.pkt.htlc.(type) {
+	case *lnwire.UpdateAddHTLC:
+		paymentHash = m.PaymentHash
+	case *lnwire.UpdateFufillHTLC, *lnwire.UpdateFailHTLC:
+		paymentHash = command.pkt.payHash
+	default:
+		command.err <- errors.New("wrong type of update")
+		return
+	}
+
+	if payment, ok := s.pendingPayments[paymentHash]; ok {
+		command.err <- s.handleUserPacket(payment, command.pkt)
+	} else {
+		command.err <- s.handleChannelLinkPacket(command.pkt)
```

Comment:
Looks like this'll cause the caller to block indefinitely as the preimage attribute isn't sent upon.